File: typescript/tests/core/fusion-set.test.ts
import { describe, it, expect } from 'vitest';
import { place } from '../../src/core/place.js';
import type { Place } from '../../src/core/place.js';
import { FusionSet, FusionSetBuilder } from '../../src/core/fusion-set.js';
import { __FUSION_SET_KEY_FOR_TEST } from '../../src/core/fusion-set.js';

/**
 * Unit tests for {@link FusionSet} per **MOD-060** (fusion set declaration).
 *
 * These tests exercise the construction and validation surface of the carrier
 * object only — the actual fusion-resolution semantics (substitution at
 * `build()` time) are covered by `fusion.test.ts`.
 *
 * Mirrors `java/src/test/java/org/libpetri/core/FusionSetTest.java`. The
 * TypeScript token-type homogeneity check is omitted at runtime (per the
 * TS-specific MOD-022 clause); compile-time `<T>` enforcement on
 * {@link FusionSetBuilder.member} and {@link FusionSet.of} provides the same
 * guarantee at the call site.
 */

describe('FusionSet — construction and validation (MOD-060)', () => {
  // ============================================================
  //  Canonical-member convention
  // ============================================================

  it('firstMemberIsCanonical: first declared member surfaces via .canonical', () => {
    const a = place<string>('A');
    const b = place<string>('B');
    const c = place<string>('C');

    const set = FusionSet.builder('limiter')
      .member(a)
      .member(b)
      .member(c)
      .build();

    expect(set.canonical).toBe(a);
    expect(set.members.length).toBe(3);
    expect(set.members[0]).toBe(a);
    expect(set.members[1]).toBe(b);
    expect(set.members[2]).toBe(c);
    expect(set.name).toBe('limiter');
  });

  // ============================================================
  //  Empty-set rejection (MOD-060)
  // ============================================================

  it('emptySet_throws: building with zero members is rejected', () => {
    expect(() => FusionSet.builder('empty').build()).toThrowError(/empty/);
  });

  // ============================================================
  //  Single-member set: degenerate but allowed (no-op fusion)
  // ============================================================

  it('singleMember_isValid: single-member fusion set is degenerate but accepted', () => {
    const only = place<string>('only');

    const set = FusionSet.builder('solo').member(only).build();

    expect(set.canonical).toBe(only);
    expect(set.members.length).toBe(1);
    expect(set.nonCanonical().length).toBe(0);
  });

  // ============================================================
  //  Static of(...) factory convenience (MOD-060)
  // ============================================================

  it('of_factoryConvenience: of(name, first, ...rest) builds the same shape', () => {
    const a = place<string>('A');
    const b = place<string>('B');
    const c = place<string>('C');

    const set = FusionSet.of('rate', a, b, c);

    expect(set.canonical).toBe(a);
    expect(set.members.length).toBe(3);
    expect(set.nonCanonical().length).toBe(2);
    expect(set.nonCanonical()[0]).toBe(b);
    expect(set.nonCanonical()[1]).toBe(c);
  });

  // ============================================================
  //  nonCanonical() returns all members except the first
  // ============================================================

  it('nonCanonical_returnsAllButFirst: the canonical is excluded, order preserved', () => {
    const a = place<number>('A');
    const b = place<number>('B');
    const c = place<number>('C');
    const d = place<number>('D');

    const set = FusionSet.builder('ints')
      .member(a)
      .member(b)
      .member(c)
      .member(d)
      .build();

    const nc = set.nonCanonical();
    expect(nc.length).toBe(3);
    expect(nc[0]).toBe(b);
    expect(nc[1]).toBe(c);
    expect(nc[2]).toBe(d);
    // canonical not in non-canonical set
    expect(nc.includes(a)).toBe(false);

    // Returned array view is structurally read-only (frozen members + slice).
    // Attempting to mutate `members` directly should fail in strict mode.
    expect(() => {
      (set.members as Place<unknown>[]).push(a);
    }).toThrow();
  });

  // ============================================================
  //  Symbol-guarded constructor
  // ============================================================

  it('constructor_isSymbolGuarded: direct construction with the wrong key is rejected', () => {
    const a = place<string>('A');
    expect(() => new FusionSet(Symbol('not-the-key'), 'x', [a])).toThrowError(
      /Use FusionSet/,
    );
  });

  it('constructor_isSymbolGuarded: construction with the right (internal) key works', () => {
    const a = place<string>('A');
    const set = new FusionSet(__FUSION_SET_KEY_FOR_TEST, 'x', [a]);
    expect(set.canonical).toBe(a);
    expect(set.members.length).toBe(1);
  });

  // ============================================================
  //  Builder validation
  // ============================================================

  it('builder_rejectsEmptyName: empty string name is rejected eagerly', () => {
    expect(() => new FusionSetBuilder('')).toThrowError(/non-empty/);
  });

  it('toString_isReadable: the toString form names canonical and member count', () => {
    const a = place<string>('A');
    const b = place<string>('B');
    const set = FusionSet.of('rate', a, b);
    expect(set.toString()).toContain('rate');
    expect(set.toString()).toContain('A');
    expect(set.toString()).toContain('2');
  });
});

File: java/src/test/java/org/libpetri/core/FusionSetTest.java
package org.libpetri.core;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link FusionSet} per <b>MOD-060</b> (fusion set declaration).
 *
 * <p>These tests exercise the construction and validation surface of the
 * carrier object only — the actual fusion-resolution semantics
 * (substitution at {@code build()} time) are covered by {@code FusionTest}.
 */
class FusionSetTest {

    // ============================================================
    //  Canonical-member convention
    // ============================================================

    @Test
    void fusionSet_firstMemberIsCanonical() {
        var a = Place.of("A", String.class);
        var b = Place.of("B", String.class);
        var c = Place.of("C", String.class);

        var set = FusionSet.builder("limiter")
            .member(a)
            .member(b)
            .member(c)
            .build();

        assertSame(a, set.canonical(),
            "first declared member must be canonical (MOD-060 convention)");
        assertEquals(3, set.members().size());
        assertEquals(a, set.members().get(0));
    }

    // ============================================================
    //  Type homogeneity (MOD-060)
    // ============================================================

    @Test
    void fusionSet_typeHomogeneity_enforced() {
        // Mixing Place<String> and Place<Integer> must fail at build().
        // We exercise the runtime check on the Builder which accepts Place<?>
        // members one at a time — the mix is created via the wildcard surface.
        var s1 = Place.of("S1", String.class);
        var s2 = Place.of("S2", String.class);
        var i1 = Place.of("I1", Integer.class);

        var builder = FusionSet.builder("mixed")
            .member(s1)
            .member(s2)
            .member(i1);

        var ex = assertThrows(IllegalArgumentException.class, builder::build);
        assertTrue(ex.getMessage().contains("I1"),
            "exception message must name the offending member. Got: " + ex.getMessage());
        assertTrue(ex.getMessage().contains("String"),
            "exception message must name the expected type. Got: " + ex.getMessage());
        assertTrue(ex.getMessage().contains("Integer"),
            "exception message must name the actual type. Got: " + ex.getMessage());
    }

    // ============================================================
    //  Empty-set rejection (MOD-060)
    // ============================================================

    @Test
    void fusionSet_emptySet_throws() {
        var ex = assertThrows(IllegalArgumentException.class,
            () -> FusionSet.builder("empty").build());
        assertTrue(ex.getMessage().contains("empty"),
            "exception message must name the empty set. Got: " + ex.getMessage());
    }

    // ============================================================
    //  Single-member set: degenerate but allowed (no-op fusion)
    // ============================================================

    @Test
    void fusionSet_singleMember_isValid() {
        var only = Place.of("only", String.class);

        var set = FusionSet.builder("solo").member(only).build();

        assertSame(only, set.canonical());
        assertEquals(1, set.members().size());
        assertTrue(set.nonCanonical().isEmpty(),
            "single-member fusion sets have no non-canonical members");
        assertEquals(String.class, set.tokenType());
    }

    // ============================================================
    //  Static of(...) factory convenience (MOD-060)
    // ============================================================

    @Test
    void fusionSet_of_factoryConvenience() {
        Place<String> a = Place.of("A", String.class);
        Place<String> b = Place.of("B", String.class);
        Place<String> c = Place.of("C", String.class);

        var set = FusionSet.of("rate", a, b, c);

        assertSame(a, set.canonical());
        assertEquals(3, set.members().size());
        assertEquals(2, set.nonCanonical().size());
        assertEquals(b, set.nonCanonical().get(0));
        assertEquals(c, set.nonCanonical().get(1));
        assertEquals(String.class, set.tokenType());
    }

    // ============================================================
    //  nonCanonical() returns all members except the first
    // ============================================================

    @Test
    void fusionSet_nonCanonical_returnsAllButFirst() {
        var a = Place.of("A", Integer.class);
        var b = Place.of("B", Integer.class);
        var c = Place.of("C", Integer.class);
        var d = Place.of("D", Integer.class);

        var set = FusionSet.builder("ints")
            .member(a)
            .member(b)
            .member(c)
            .member(d)
            .build();

        var nonCanonical = set.nonCanonical();
        assertEquals(3, nonCanonical.size());
        assertEquals(b, nonCanonical.get(0));
        assertEquals(c, nonCanonical.get(1));
        assertEquals(d, nonCanonical.get(2));
        // canonical not in non-canonical set
        assertFalse(nonCanonical.contains(a));

        // Verify nonCanonical is a list view, not modifiable.
        assertThrows(UnsupportedOperationException.class,
            () -> nonCanonical.add(a));
    }
}
